### PR TITLE
add rust-based stage1 driver

### PR DESCRIPTION
**Pull request by Mic92**

### CODE REVIEW

**Comment by Mic92** [comment]
- replaces insmod virtio with stage1 kernel driver
- replaces niv with nix flakes (requires nixUnstable)
- adds rustc for stage1 kernel driver and fenix

**Comment by Mic92** [comment]
I still need to fix the virtio-blk test so and set KERNELDIR in ubuntu ci.

File: src/tracer/wrap_syscall.rs [hunk 1]
```diff
@@ -323,7 +323,15 @@ impl KvmRunWrapper {
                 return self.stopped(pid, &signal);
             }
             WaitStatus::Exited(pid, status) => {
-                bail!("process {} exited with: {}", pid, status)
+                if let Some(index) = self
+                    .threads
+                    .iter()
+                    .position(|thread| thread.ptthread.tid == pid)
+                {
+                    self.threads.remove(index);
```
Comment (pogobanane) [inline src/tracer/wrap_syscall.rs:331]:
```
        // remove and shift others to left
        self.threads.remove(idx);
        // shift idx if it was shifted
        if idx > self.process_idx {
            self.process_idx -= 1;
        }
```

File: src/tracer/wrap_syscall.rs [hunk 2]
```diff
@@ -322,7 +322,17 @@ impl KvmRunWrapper {
...
                 return self.stopped(pid, &signal);
             }
-            WaitStatus::Exited(_, status) => bail!("process exited with: {}", status),
+            WaitStatus::Exited(pid, status) => {
+                if let Some(index) = self
+                    .threads
+                    .iter()
+                    .position(|thread| thread.ptthread.tid == pid)
+                {
+                    self.threads.remove(index);
```
Comment (Mic92) [inline src/tracer/wrap_syscall.rs:331]: Do you mean? 
```suggestion
                    // remove and shift others to left
                    self.threads.remove(idx);
                    // shift idx if it was shifted
                    if idx > self.process_idx {
                        self.process_idx -= 1;
                    }

```

File: src/tracer/wrap_syscall.rs (hunk 1 above)
Comment (pogobanane) [reply src/tracer/wrap_syscall.rs:331]: no its `idx < self.process_idx`

File: src/tracer/wrap_syscall.rs (hunk 2 above)
Comment (pogobanane) [reply src/tracer/wrap_syscall.rs:331]: yes (`>` -> `<`)
```suggestion
                    // remove and shift others to left
                    self.threads.remove(idx);
                    // shift idx if it was shifted
                    if idx < self.process_idx {
                        self.process_idx -= 1;
                    }
```

Comment (pogobanane) [reply src/tracer/wrap_syscall.rs:331]: it fixes the crash but it looks like it leaves the device unresponsive. Not sure if this issue is related.

**Pull request merged**